### ISSUE TITLE
mingw-w64: Per subproject default_library doesn't work well for runtime dependencies

### ISSUE THREAD

**Issue by pal1000** [mesonbuild/meson#7785]
Let's assume you have a meson project that has at least 1 dependency that is prebuilt with a mingw-w64 toolchain and it's somehow automatically discoverable using pkg-config method (it's a runtime dependency). Obviously you attempt to build main project with same mingw toolchain, but you want the dependenc(y/ies) to be statically linked and you use `-Dsubproject:default_library=static` to enforce it.

**Reproducer**
As example let's try building Mesa3D softpipe in MSYS2
- Open a MSYS2 MinGW 32-bit or 64-bit shell, this determines if build is 64-bit or 32-bit;
- Install build dependencies: `pacman -S ${MINGW_PACKAGE_PREFIX}-meson  ${MINGW_PACKAGE_PREFIX}-pkg-config  ${MINGW_PACKAGE_PREFIX}-gcc  ${MINGW_PACKAGE_PREFIX}-zlib flex bison --needed`
- Build configure: `meson build  -Dzlib:default_library=static --buildtype=release -Dc_link_args='-static' -Dcpp_link_args='-static' --prefix=...`.

**Expected behavior**
Resulted `opengl32.dll` file in `build\src\gallium\targets\libgl-gdi` should be portable and not depend on any MSYS2 MinGW toolchain or zlib dls, Unfortunately it depends on `zlib1.dll`. It appears `-Dzlib:default_library=static` fails somehow. `-Dc_link_args='-static' -Dcpp_link_args='-static'` combo acts like equivalent for `-Db_vsscrt=mt` as expected.

**System parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? Native build
* what operating system : Windows 10
* what Python version are you using : 3.8.5
* what `meson --version` : 0.55.2
* what `ninja --version` : 1.10.1

See also msys2/MINGW-packages#6993

**Comment by xclaesse** [comment]
I think there is a misunderstanding. default_library=static means your project will build static libraries, but does not mean that dependencies must be static too. You want to lookup your pkgconfig dependencies with `dependency('foo', static: true)`.

**Comment by pal1000** [comment]
@xclaesse I meant `-Dsubproject:default_library` where `subproject` is name of subproject. 
I expected `-Dsubproject:default_library=static` to make `dependency('subproject')` identical to `dependency('subproject', static: true)`.

Changed title to make it clear that this issue is exclusively about per subproject default_library.

**Comment by eli-schwartz** [comment]
> that has at least 1 dependency that is prebuilt with a mingw-w64 toolchain and it's somehow automatically discoverable using pkg-config method (it's a runtime dependency). 

So it's not using the subproject fallback.

> Obviously you attempt to build main project with same mingw toolchain, but you want the dependenc(y/ies) to be statically linked and you use `-Dsubproject:default_library=static` to enforce it.

But you told meson "if it gets built as a subproject, build it statically".

> I expected `-Dsubproject:default_library=static` to make `dependency('subproject')` identical to `dependency('subproject', static: true)`

It is not. It is identical to passing the default_options kwarg to subproject() or dependency() functions as such:


```
subproject('subproject', default_options: ['default_library=static'])
```

```
dependency('subproject', static: false, fallback: ['subproject', 'subproj_dep'], default_options: ['default_library=static'])
```

Hence it is called "default_library", a direct reference to the relevant project option for telling meson how to compile the result of a `library('libname', sources)` function.

...

What you *wanted* was an option `-Dstatic_dependencies=true`, but no such option exists. It is a reasonable desire, I want it too. I don't have it, so instead I have meson_options.txt:

```
option('buildstatic', type : 'boolean', value : false,
       description : 'if true, build statically linked binaries')
```

and every single `dependency('foo')` is actually `dependency('foo', static : get_option('buildstatic'))` which is ugly and inelegant, but also works.

What isn't a reasonable desire: wanting the UI for passing meson subproject options to also affect system dependency lookup.

**Comment by eli-schwartz** [comment]
As such, this issue should be closed as a duplicate of #7621

**Comment by pal1000** [comment]
Ok, @ eli-schwartz
